File: rerun-reports/lane-a/070/20260828-120814/results/storage/data-plane/js-ts/crud/storage-dp-js-ts-encrypted-uploader/js-ts-azure-skills/azure-skill-mcp-microsoft-skill/generated-code/src/keyManagement.ts
import { randomBytes } from "node:crypto";
import type { TokenCredential } from "@azure/core-auth";
import {
  CryptographyClient,
  type KeyClient,
} from "@azure/keyvault-keys";

const DATA_KEY_BYTES = 32;
export const KEY_WRAP_ALGORITHM = "RSA-OAEP-256" as const;

export interface ProtectedDataKey {
  readonly dataKey: Buffer;
  readonly keyId: string;
  readonly wrappedKey: Buffer;
  readonly wrapAlgorithm: typeof KEY_WRAP_ALGORITHM;
}

interface AzureErrorDetails {
  readonly code?: string;
  readonly message?: string;
  readonly statusCode?: number;
}

function getAzureErrorDetails(error: unknown): AzureErrorDetails {
  if (typeof error !== "object" || error === null) {
    return {};
  }

  const candidate = error as Record<string, unknown>;
  return {
    ...(typeof candidate.code === "string" ? { code: candidate.code } : {}),
    ...(typeof candidate.message === "string"
      ? { message: candidate.message }
      : {}),
    ...(typeof candidate.statusCode === "number"
      ? { statusCode: candidate.statusCode }
      : {}),
  };
}

export class KeyManagementError extends Error {
  public constructor(operation: string, cause: unknown) {
    const details = getAzureErrorDetails(cause);
    const context = [
      details.statusCode === undefined
        ? undefined
        : `status ${details.statusCode}`,
      details.code,
    ]
      .filter((value): value is string => value !== undefined)
      .join(", ");
    const reason = details.message ? `: ${details.message}` : "";

    super(
      `Azure Key Vault ${operation} failed${context ? ` (${context})` : ""}${reason}`,
      { cause },
    );
    this.name = "KeyManagementError";
  }
}

export class KeyManagement {
  public constructor(
    private readonly keyClient: KeyClient,
    private readonly credential: TokenCredential,
    private readonly vaultUrl: string,
    private readonly keyName: string,
  ) {}

  public async createProtectedDataKey(): Promise<ProtectedDataKey> {
    const dataKey = randomBytes(DATA_KEY_BYTES);

    try {
      const key = await this.keyClient.getKey(this.keyName);
      if (!key.id) {
        throw new Error(`Key Vault returned no ID for key ${this.keyName}.`);
      }

      const cryptographyClient = new CryptographyClient(
        key.id,
        this.credential,
      );
      const wrapResult = await cryptographyClient.wrapKey(
        KEY_WRAP_ALGORITHM,
        dataKey,
      );

      return {
        dataKey,
        keyId: key.id,
        wrappedKey: Buffer.from(wrapResult.result),
        wrapAlgorithm: KEY_WRAP_ALGORITHM,
      };
    } catch (error) {
      dataKey.fill(0);
      if (error instanceof KeyManagementError) {
        throw error;
      }
      throw new KeyManagementError("key lookup or wrap", error);
    }
  }

  public async recoverDataKey(
    keyId: string,
    wrappedKey: Buffer,
    wrapAlgorithm: string,
  ): Promise<Buffer> {
    if (wrapAlgorithm !== KEY_WRAP_ALGORITHM) {
      throw new KeyManagementError(
        "unwrap",
        new Error(`Unsupported key wrap algorithm: ${wrapAlgorithm}`),
      );
    }

    this.assertConfiguredKeyId(keyId);

    try {
      const cryptographyClient = new CryptographyClient(keyId, this.credential);
      const unwrapResult = await cryptographyClient.unwrapKey(
        KEY_WRAP_ALGORITHM,
        wrappedKey,
      );
      const dataKey = Buffer.from(unwrapResult.result);

      if (dataKey.length !== DATA_KEY_BYTES) {
        dataKey.fill(0);
        throw new Error(
          `Key Vault returned a ${dataKey.length}-byte data key; expected ${DATA_KEY_BYTES} bytes.`,
        );
      }

      return dataKey;
    } catch (error) {
      if (error instanceof KeyManagementError) {
        throw error;
      }
      throw new KeyManagementError("unwrap", error);
    }
  }

  private assertConfiguredKeyId(keyId: string): void {
    let configuredVault: URL;
    let candidate: URL;

    try {
      configuredVault = new URL(this.vaultUrl);
      candidate = new URL(keyId);
    } catch (error) {
      throw new KeyManagementError("key ID validation", error);
    }

    const pathParts = candidate.pathname.split("/").filter(Boolean);
    const hasExpectedPath =
      pathParts.length === 3 &&
      pathParts[0]?.toLowerCase() === "keys" &&
      pathParts[1] === this.keyName &&
      Boolean(pathParts[2]);

    if (
      candidate.protocol !== "https:" ||
      candidate.origin !== configuredVault.origin ||
      candidate.search ||
      candidate.hash ||
      !hasExpectedPath
    ) {
      throw new KeyManagementError(
        "key ID validation",
        new Error(
          "Blob metadata references a key outside the configured vault, key name, or key version.",
        ),
      );
    }
  }
}

File: rerun-reports/lane-a/041/20260828-050611/results/storage/data-plane/java/crud/storage-dp-java-encrypted-uploader/java-azure-skills/azure-skill-mcp-microsoft-skill/generated-code/src/main/java/com/example/encryptedblob/KeyManagement.java
package com.example.encryptedblob;

import com.azure.core.credential.TokenCredential;
import com.azure.core.exception.HttpResponseException;
import com.azure.security.keyvault.keys.KeyClient;
import com.azure.security.keyvault.keys.cryptography.CryptographyClient;
import com.azure.security.keyvault.keys.cryptography.CryptographyClientBuilder;
import com.azure.security.keyvault.keys.cryptography.models.KeyWrapAlgorithm;
import com.azure.security.keyvault.keys.models.KeyVaultKey;

import java.security.SecureRandom;
import java.util.Arrays;

public final class KeyManagement {
    private static final int DATA_KEY_BYTES = 32;
    private static final KeyWrapAlgorithm WRAP_ALGORITHM = KeyWrapAlgorithm.RSA_OAEP_256;

    private final KeyClient keyClient;
    private final TokenCredential credential;
    private final String keyName;
    private final SecureRandom secureRandom;

    public KeyManagement(KeyClient keyClient, TokenCredential credential, String keyName) {
        this.keyClient = keyClient;
        this.credential = credential;
        this.keyName = keyName;
        this.secureRandom = new SecureRandom();
    }

    DataKeyEnvelope generateAndWrapKey() {
        byte[] dataKey = new byte[DATA_KEY_BYTES];
        secureRandom.nextBytes(dataKey);

        try {
            KeyVaultKey vaultKey = keyClient.getKey(keyName);
            String versionedKeyId = vaultKey.getId();
            CryptographyClient cryptographyClient = cryptoClient(versionedKeyId);
            byte[] wrappedKey = cryptographyClient.wrapKey(WRAP_ALGORITHM, dataKey).getEncryptedKey();
            return new DataKeyEnvelope(dataKey, new ProtectedDataKey(versionedKeyId, wrappedKey));
        } catch (HttpResponseException exception) {
            Arrays.fill(dataKey, (byte) 0);
            throw keyVaultFailure("wrap a data encryption key", exception);
        } catch (RuntimeException exception) {
            Arrays.fill(dataKey, (byte) 0);
            throw exception;
        }
    }

    byte[] unwrapKey(ProtectedDataKey protectedKey) {
        try {
            return cryptoClient(protectedKey.keyId())
                .unwrapKey(WRAP_ALGORITHM, protectedKey.wrappedKey())
                .getKey();
        } catch (HttpResponseException exception) {
            throw keyVaultFailure("unwrap the data encryption key", exception);
        }
    }

    private CryptographyClient cryptoClient(String keyId) {
        return new CryptographyClientBuilder()
            .keyIdentifier(keyId)
            .credential(credential)
            .buildClient();
    }

    private static EnvelopeEncryptionException keyVaultFailure(String operation, HttpResponseException exception) {
        int status = exception.getResponse() == null ? -1 : exception.getResponse().getStatusCode();
        return new EnvelopeEncryptionException(
            "Key Vault could not " + operation + " (HTTP " + status
                + "). Verify that the stored key version is enabled and permits wrapKey/unwrapKey.",
            exception);
    }
}
